tasks.py

```diff
@@ -231,6 +233,91 @@ def dev_env(ctx, architecture="amd64", name="kind", cni=None):
         tmp.flush()
         run("kubectl apply -f {}".format(tmp.name), echo=True)
 
+    if protocol == "bgp":
+        print("Configuring MetalLB with a BGP test environment")
+        bgp_dev_env()
+    else:
+        print("Leaving MetalLB unconfigured")
+
+
+# Configure MetalLB in the dev-env for BGP testing.  Start an frr based BGP
+# router in a container and configure MetalLB to peer with it.
+# See dev-env/bgp/README.md for some more information.
+def bgp_dev_env():
+    dev_env_dir = os.getcwd() + "/dev-env/bgp"
+    frr_volume_dir = dev_env_dir + "/frr-volume"
+
+    # TODO -- The IP address handling will need updates to add support for IPv6
+
+    # We need the IPs for each Node in the cluster to place them in the BGP
+    # router configuration file (bgpd.conf).  Each Node will peer with this
+    # router.
+    node_ips = run("kubectl get nodes -o jsonpath='{.items[*].status.addresses"
+            "[?(@.type==\"InternalIP\")].address}'", echo=True)
+    node_ips = node_ips.stdout.strip().split()
+    if len(node_ips) != 3:
+        raise Exit(message='Expected 3 nodes, got %d' % len(node_ips))
+
+    # Create a new directory that will be used as the config volume for frr.
+    try:
+        # sudo because past docker runs will have changed ownership of this dir
+        sudo('rm -rf "%s"' % frr_volume_dir)
+        os.mkdir(frr_volume_dir)
+    except FileExistsError:
+        pass
+    except Exception as e:
+        raise Exit(message='Failed to create frr-volume directory: %s'
+                   % str(e))
+
+    # These two config files are static, so we just copy them straight in.
+    shutil.copyfile("%s/frr/zebra.conf" % dev_env_dir,
+                    "%s/zebra.conf" % frr_volume_dir)
+    shutil.copyfile("%s/frr/daemons" % dev_env_dir,
+                    "%s/daemons" % frr_volume_dir)
+
+    # bgpd.conf is created from a template so that we can include the current
+    # Node IPs.
+    with open("%s/frr/bgpd.conf.tmpl" % dev_env_dir, 'r') as f:
+        bgpd_config = "! THIS FILE IS AUTOGENERATED\n" + f.read()
+    for n in range(0, len(node_ips)):
+        bgpd_config = bgpd_config.replace("NODE%d_IP" % n, node_ips[n])
+    with open("%s/bgpd.conf" % frr_volume_dir, 'w') as f:
+        f.write(bgpd_config)
+
+    # Run a BGP router in a container for all of the speakers to peer with.
+    run('([ "$(docker ps -a | grep frr)" ] && docker rm -f frr) || /bin/true')
```

Comment:
We probably want something like: `docker ps -a -f name=frr` instead of the grep. If someone has been playing with frr (like me :D) the grep might return something but the docker rm later will fail

Reply:
Good point, I will adjust

Reply:
done